### PR TITLE
fix(s3): InventoryConfiguration[].Id exceeds maximum size allowed by the PutBucketInventoryConfiguration API

### LINKED ISSUE #27793: (@aws-cdk/s3): InventoryConfiguration[].Id exceeds maximum size allowed by the PutBucketInventoryConfiguration API, when NodeId is overly long
### Describe the bug

In cases where the node-id of the bucket is quite long, the generated `InventoryConfiguration[].Id` will be longer than the [S3 PutBucketInventoryConfiguration](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketInventoryConfiguration.html) API allows. This will cause the cloudformation deployment to fail:

```
Service: Amazon S3; Status Code: 400; Error Code: MalformedXML; Request ID: ...; S3 Extended Request ID: ...; Proxy: null), The XML you provided was not well-formed or did not validate against our published schema.
```

### Expected Behavior

The automatically generated id will satisfy the S3 API and the deployment will be successful.

### Current Behavior

The deployment fails and Cloudformation errors:

```
2:09:29 PM | CREATE_FAILED        | AWS::S3::Bucket       | VeryVeryVeryVeryVe...NodeIdName0592D76A
The XML you provided was not well-formed or did not validate against our published schema (Service: Amazon S3; Status Code: 400; Error Code: MalformedXML; Request ID: RKRENTWF25HGT99P; S3 Extended Request ID: CB8ezYmFdgU+KE9ZABh1sVEp4v
U1kd7r3Kdkse/pxrNTQImcbrFoO7ODJlwVsVRDTcwbLZIRLE8=; Proxy: null)
```

### Reproduction Steps

Example to reproduce:

```ts
const stack = new cdk.Stack();
  
  const inventoryBucket = new s3.Bucket(stack, 'InventoryBucket');
  new s3.Bucket(stack, 'VeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongNodeIdName', {
    inventories: [
      {
        destination: {
          bucket: inventoryBucket,
        },
      },
    ],
  });
```

### Possible Solution

Truncating the inventory-id to 64 characters.

---

A workaraound is to set the id manually:

```ts
const stack = new cdk.Stack();

const inventoryBucket = new s3.Bucket(stack, 'InventoryBucket');
new s3.Bucket(stack, 'VeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVery&LongNodeIdName', {
  inventories: [
    {
      destination: {
        bucket: inventoryBucket,
      },
      inventoryId: 'a-short-id'
    },
  ],
});
```

### Additional Information/Context

Relevant correspondence with the AWS support

> Hello ***, 
> 
> I hope you are doing well. This is *** again from AWS premium support.
> 
> [...]
> Our service team has identified that the inventory configuration ID should be limited to 64 characters and can only contain letters, numbers, periods, dashes, and underscores.
> 
> Upon investigation, we found that the ID generated in the shared CDK code exceeds the 64-character limit. We are actively working on updating our documentation to ensure that this information is readily available to other customers in the future. We deeply regret that it took some time to identify that the root cause of this issue was simply an extra character in the ID.
> 
> [...]
> 
> Best regards,
> \***
> Amazon Web Services

### CDK CLI Version

2.103.1 (build 3bb19ac)

### Framework Version

2.103.1

### Node.js Version

v18.17.0

### OS

Linux 5.15.133-1-MANJARO #1 SMP PREEMPT Sat Sep 23 10:10:02 UTC 2023 x86_64 GNU/Linux

### Language

TypeScript

### Language Version

Typescript 5.2.2

### Other information

_No response_

### PR DESCRIPTION
**Pull request by bweigel**
Truncate the inventory-id to 64 characters, such that it conforms to the  S3 PutBucketInventoryConfiguration API.
Also remove non-allowed characters.

Closes #27793.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by bweigel** [comment]
Exemption Request: I don't think integration test changes are needed here, but happy to add them if needed.

**Comment by bweigel** [comment]
Exemption Request: I don't think integration test changes are needed here, but happy to add them if needed.

File: packages/aws-cdk-lib/aws-s3/lib/bucket.ts [hunk 1]
```diff
@@ -2424,7 +2424,7 @@ export class Bucket extends BucketBase {
     return this.inventories.map((inventory, index) => {
       const format = inventory.format ?? InventoryFormat.CSV;
       const frequency = inventory.frequency ?? InventoryFrequency.WEEKLY;
-      const id = inventory.inventoryId ?? `${this.node.id}Inventory${index}`;
+      const id = inventory.inventoryId ?? `${this.node.id}Inventory${index}`.replace( /[^\w\.\-]/g, '').slice(-64);
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-s3/lib/bucket.ts:2427]: Are these conditions in the docs somewhere? I don't see them [here](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketInventoryConfiguration.html).

Comment (bweigel) [reply packages/aws-cdk-lib/aws-s3/lib/bucket.ts:2427]: Correct, @msambol they are not documented. That's why I added the conversation with the AWS Premium Support above.

File: packages/aws-cdk-lib/aws-s3/test/bucket.test.ts [hunk 2]
```diff
@@ -3134,6 +3134,30 @@ describe('bucket', () => {
     });
   });
 
+  test('Inventory Ids are shortened to 64 characters', () => {
+    // Given
+    const stack = new cdk.Stack();
+
+    const inventoryBucket = new s3.Bucket(stack, 'InventoryBucket');
+    new s3.Bucket(stack, 'AVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVery&LongNodeIdName', {
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3142]: can you add a test or change some characters in this one to show the unallowed characters being removed?

Comment (bweigel) [reply packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3142]: There is an & in the string. Does this suffice, or should I make it more obvious?

Comment (msambol) [reply packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3142]: Sorry, I missed that. I think a few more would make it easier to see. Thanks!

**Comment by msambol** [comment]
LGTM.

File: packages/aws-cdk-lib/aws-s3/lib/bucket.ts (hunk 1 above)
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-s3/lib/bucket.ts:2427]: If `inventoryId` is explicitly provided with bad characters / is too long, I think we should validate with a clear error message.

Can you also put an explanation in the doc string (~ line 1253).

I can understand replacing bad characters if they are taken from the node id.

File: packages/aws-cdk-lib/aws-s3/test/bucket.test.ts [hunk 3]
```diff
@@ -3134,6 +3134,30 @@ describe('bucket', () => {
...
+
+    Template.fromStack(stack).hasResourceProperties('AWS::S3::Bucket', {
+      InventoryConfigurations: Match.arrayWith([
+        Match.objectLike({
+          Id: 'VeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongNodeIdNameInventory0',
+        }),
+      ]),
+    });
+  });
+
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3160]: When you add the validation for `inventoryId`, could you add a fail test here?

**Review by scanlonp — DISMISSED** [review]
Good change, a couple small comments.

I also think we need an integ test. It can be simple, but will serve to both confirm the id is well formed as well as preventing future changes from breaking this again.

File: packages/aws-cdk-lib/aws-s3/lib/bucket.ts (hunk 1 above)
Comment (bweigel) [reply packages/aws-cdk-lib/aws-s3/lib/bucket.ts]: Done

**Comment by bweigel** [comment]
@scanlonp just a heads up, I won't be able to work on the PR until December starting Sunday.

File: packages/aws-cdk-lib/aws-s3/lib/bucket.ts [hunk 4]
```diff
@@ -2420,11 +2421,15 @@ export class Bucket extends BucketBase {
...
       return undefined;
     }
+    const inventoryIdValidationRegex = /[^\w\.\-]/g;
 
     return this.inventories.map((inventory, index) => {
       const format = inventory.format ?? InventoryFormat.CSV;
       const frequency = inventory.frequency ?? InventoryFrequency.WEEKLY;
-      const id = inventory.inventoryId ?? `${this.node.id}Inventory${index}`.replace( /[^\w\.\-]/g, '').slice(-64);
+      if (inventory.inventoryId !== undefined && (inventory.inventoryId.length > 64 || inventoryIdValidationRegex.test(inventory.inventoryId))) {
+        throw new Error('InventoryId should not exceed 64 characters and should not contain special characters except . and -');
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-s3/lib/bucket.ts:2430]:
```suggestion
        throw new Error(`inventoryId should not exceed 64 characters and should not contain special characters except . and -, got ${inventory.inventoryId}`);
```

File: packages/aws-cdk-lib/aws-s3/test/bucket.test.ts [hunk 5]
```diff
@@ -3158,6 +3158,25 @@ describe('bucket', () => {
...
+        {
+          destination: {
+            bucket: inventoryBucket,
+          },
+          inventoryId: 'InvalidId&/',
+        },
+      ],
+    });
+
+    expect(() => Template.fromStack(stack)).toThrow(/InventoryId should not exceed 64 characters and should not contain special characters except . and -/);
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3177]:
```suggestion
    expect(() => Template.fromStack(stack)).toThrow(/inventoryId should not exceed 64 characters and should not contain special characters except . and -, got InvalidId&123/);
```

File: packages/aws-cdk-lib/aws-s3/test/bucket.test.ts [hunk 6]
```diff
@@ -3134,6 +3134,49 @@ describe('bucket', () => {
...
+    const stack = new cdk.Stack();
+
+    const inventoryBucket = new s3.Bucket(stack, 'InventoryBucket');
+    new s3.Bucket(stack, 'MyBucket2', {
+      inventories: [
+        {
+          destination: {
+            bucket: inventoryBucket,
+          },
+          inventoryId: 'InvalidId&/',
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-s3/test/bucket.test.ts:3172]:
```suggestion
          inventoryId: 'InvalidId&123',
```

**Review by scanlonp — COMMENTED** [review]
This looks very good! Last small ask is to return the value in the error message.

**Review by scanlonp — DISMISSED** [review]
Good to go if the build passes!

**Review by scanlonp — APPROVED** [review]

**Pull request merged**